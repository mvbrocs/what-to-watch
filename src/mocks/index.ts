export { default as mockFilms } from './films';
export { default as mockStore } from './store';
