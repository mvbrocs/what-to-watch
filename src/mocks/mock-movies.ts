import img1 from './img/fantastic-beasts-the-crimes-of-grindelwald.jpg';
import img2 from './img/bohemian-rhapsody.jpg';
import img3 from './img/macbeth.jpg';
import img4 from './img/aviator.jpg';
import img5 from './img/we-need-to-talk-about-kevin.jpg';

export interface IMovie {
  id: string;
  title: string;
  genre: string;
  img: {
    src: string;
    alt: string;
    width: number;
    height: number;
  };
  videoSrc: string;
}

export const mockMovies: IMovie[] = [
  {
    id: 'movie-1',
    title: 'Fantastic Beasts: The Crimes of Grindelwald',
    genre: 'Fantastic',
    img: {
      src: img1,
      alt: 'Fantastic Beasts: The Crimes of Grindelwald',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4',
  },
  {
    id: 'movie-2',
    title: 'Bohemian Rhapsody',
    genre: 'Dramas',
    img: {
      src: img2,
      alt: 'Bohemian Rhapsody',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm',
  },
  {
    id: 'movie-3',
    title: 'Macbeth',
    genre: 'Comedy',
    img: {
      src: img3,
      alt: 'Macbeth',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4',
  },
  {
    id: 'movie-4',
    title: 'Aviator',
    genre: 'Horror',
    img: {
      src: img4,
      alt: 'Aviator',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm',
  },
  {
    id: 'movie-5',
    title: 'What We Do in the Shadows',
    genre: 'Crime',
    img: {
      src: img5,
      alt: 'What We Do in the Shadows',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4',
  },
];
