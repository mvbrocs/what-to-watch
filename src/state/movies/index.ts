import { moviesSlice } from './movies-slice';

export const moviesReducer = moviesSlice.reducer;
export { fetchMovies } from './movies-slice';
export const { incrementMaxVisibleMovies } = moviesSlice.actions;
export {
  selectMoviesSlice,
  selectAllGenres,
  selectMoviesByGenre,
  selectAllMoviesIsVisible,
  selectMoviesByGenreAndMaxVisible,
} from './movies-selectors';
