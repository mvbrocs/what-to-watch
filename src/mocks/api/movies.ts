import { IMovie, movies } from 'mocks/movies';

export const moviesAPI = {
  getAll: () =>
    new Promise<{ data: IMovie[] }>((resolve) => {
      setTimeout(() => resolve({ data: movies }), 2000);
    }),
};
