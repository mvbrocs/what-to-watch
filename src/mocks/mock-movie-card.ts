import { mockMovies } from './mock-movies';

export const mockMovieCard = mockMovies[0];
