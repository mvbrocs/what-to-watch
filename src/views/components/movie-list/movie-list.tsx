import React from 'react';

import { MovieCard } from 'views/components/movie-card/movie-card';
import { IMovie } from 'mocks/mock-movies';

type MovieListProps = {
  movies: IMovie[];
};

export const MovieList = ({ movies }: MovieListProps) => (
  <div className="catalog__movies-list">
    {movies.map((movie) => (
      <MovieCard {...movie} key={movie.id} />
    ))}
  </div>
);
