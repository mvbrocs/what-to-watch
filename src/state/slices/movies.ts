import {
  createAsyncThunk,
  createEntityAdapter,
  createSelector,
  createSlice,
  EntityState,
} from '@reduxjs/toolkit';

import { ALL_GENRES } from 'src/state/slices/ui';
import { api } from 'src/api';
import { IFilm } from 'src/api/films';
import { RootState } from 'src/state/root-reducer';
import { AppThunk } from 'src/state/store';

type State = {
  loaded: boolean;
  loading: boolean;
  didInvalidate: boolean;
  error: string | null;
} & EntityState<IFilm>;

const fetchMovies = createAsyncThunk('movies/getAll', async () => {
  const { data } = await api.films.getAll();

  return data;
});

const shouldFetchMovies = (state: RootState) => {
  const movies = state.movies;

  if (!movies.loaded) return true;

  if (movies.loading) return false;

  return movies.didInvalidate;
};

export const fetchMoviesIfNeeded = (): AppThunk => (dispatch, getState) => {
  if (shouldFetchMovies(getState())) return dispatch(fetchMovies());

  return Promise.resolve();
};

export const moviesInitialState = {
  loaded: false,
  loading: false,
  didInvalidate: false,
  error: null,
} as State;

export const moviesAdapter = createEntityAdapter<IFilm>();

const slice = createSlice({
  name: 'movies',
  initialState: moviesAdapter.getInitialState(moviesInitialState),
  reducers: {
    invalidate(state) {
      state.didInvalidate = true;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchMovies.pending, (state: State) => {
      state.loaded = false;
      state.loading = true;
      state.error = null;
      state.didInvalidate = false;
      moviesAdapter.setAll(state, []);
    });
    builder.addCase(fetchMovies.fulfilled, (state: State, action) => {
      state.loaded = true;
      state.loading = false;
      moviesAdapter.setAll(state, action.payload);
    });
  },
});

export const moviesReducer = slice.reducer;

export const { invalidate: invalidateMovies } = slice.actions;

export const selectMoviesSlice = (state: RootState) => state.movies;

export const {
  selectAll: selectAllMovies,
  selectById: selectMovieByID,
} = moviesAdapter.getSelectors(selectMoviesSlice);

export const selectAllGenres = createSelector(selectAllMovies, (movies) => {
  const genres = [ALL_GENRES];

  for (const movie of movies) if (!genres.includes(movie.genre)) genres.push(movie.genre);

  return genres;
});
