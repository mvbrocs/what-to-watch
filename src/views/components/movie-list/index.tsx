import React from 'react';

import { MovieCard } from 'views/components/movie-card';
import { IFilm } from 'api/films';

type MovieListProps = {
  movies: IFilm[];
};

export const MovieList = ({ movies }: MovieListProps) => (
  <div className="catalog__movies-list">
    {movies.map((movie) => (
      <MovieCard {...movie} key={movie.id} />
    ))}
  </div>
);
