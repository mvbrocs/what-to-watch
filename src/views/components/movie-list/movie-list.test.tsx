import React from 'react';
import { shallow } from 'enzyme';

import { MovieList } from './movie-list';
import { movies } from 'mocks/movies';

describe('MovieList', () => {
  const movieList = shallow(<MovieList movies={movies} />);

  it('renders properly', () => {
    expect(movieList).toMatchSnapshot();
  });
});
