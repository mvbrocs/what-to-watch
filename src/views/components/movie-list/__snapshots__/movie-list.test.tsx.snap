// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList renders properly 1`] = `
<div
  className="catalog__movies-list"
>
  <MovieCard
    id="movie-1"
    img={
      Object {
        "alt": "Fantastic Beasts: The Crimes of Grindelwald",
        "height": 175,
        "src": "fantastic-beasts-the-crimes-of-grindelwald.jpg",
        "width": 280,
      }
    }
    key="movie-1"
    title="Fantastic Beasts: The Crimes of Grindelwald"
    videoSrc="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
  />
  <MovieCard
    id="movie-2"
    img={
      Object {
        "alt": "Bohemian Rhapsody",
        "height": 175,
        "src": "bohemian-rhapsody.jpg",
        "width": 280,
      }
    }
    key="movie-2"
    title="Bohemian Rhapsody"
    videoSrc="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
  />
  <MovieCard
    id="movie-3"
    img={
      Object {
        "alt": "Macbeth",
        "height": 175,
        "src": "macbeth.jpg",
        "width": 280,
      }
    }
    key="movie-3"
    title="Macbeth"
    videoSrc="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
  />
  <MovieCard
    id="movie-4"
    img={
      Object {
        "alt": "Aviator",
        "height": 175,
        "src": "aviator.jpg",
        "width": 280,
      }
    }
    key="movie-4"
    title="Aviator"
    videoSrc="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
  />
  <MovieCard
    id="movie-5"
    img={
      Object {
        "alt": "What We Do in the Shadows",
        "height": 175,
        "src": "we-need-to-talk-about-kevin.jpg",
        "width": 280,
      }
    }
    key="movie-5"
    title="What We Do in the Shadows"
    videoSrc="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
  />
</div>
`;
